src/Data/List/Properties.agda

```diff
@@ -575,6 +576,43 @@ module _ {a} {A : Set a} where
   tabulate-lookup []       = refl
   tabulate-lookup (x ∷ xs) = P.cong (_ ∷_) (tabulate-lookup xs)
 
+  tabulate-len : ∀ {n} → (f : Fin n → A) →
+                 length (tabulate f) ≡ n
+  tabulate-len {zero} f = refl
+  tabulate-len {suc n} f = cong suc (tabulate-len (λ z → f (suc z)))
+
+  tabulate-len≤l : ∀{n} → (f : Fin n → A) →
+                   length (tabulate f) ≤ n
+  tabulate-len≤l {n = zero} f = z≤n
+  tabulate-len≤l {n = suc n} f = s≤s (tabulate-len≤l (λ z → f (suc z)))
+
+  tabulate-len≤r : ∀{n} → (f : Fin n → A) →
+                   n ≤ length (tabulate f)
+  tabulate-len≤r {n = zero} f = z≤n
+  tabulate-len≤r {n = suc n} f = s≤s (tabulate-len≤r (λ z → f (suc z)))
+
+
+  lookup-tabulate : ∀{n} → (f : Fin n → A) →
+                    ∀ x → let x′ = subst Fin (tabulate-len f) x
+                          in f x′ ≡ lookup (tabulate f) x
```

Comment:
x′ can be written as `cast (length-tabulate f) x` where `case` is from `Data.Fin`. The letter `i` should be used for indices rather than `x`